#' help UI Function
#'
#' @description Shiny module containing the 'Help' section as of the UI.
#'
#' @param id,input,output,session Internal parameters for {shiny}.
#'
#' @noRd 
#'
#' @importFrom shiny NS tagList 
mod_help_ui <- function(id){
  ns <- NS(id)
  tagList(
    
    h2("Getting started with aRgus"),
    
    h3("Toolbar"),
    
    strong("Select gene"),
    p("The initial and sole mandatory input for an aRgus query is the HGNC symbol of your gene of interest which can be either
      directly typed in the search bar or prompted after entering single letters contained by the desired gene name (> ", em("Select gene"), ").
      Prompts can subsequently be selected from the drop-down list which initiates the query. In order to harmonize variant data, aRgus
      now selects the corresponding Ensembl feature ID and RefSeq ID of the ", a("MANE", href = "https://www.ncbi.nlm.nih.gov/refseq/MANE/"), 
      "-curated canonical transcript. If no MANE transcript is available, the highest curated ", a("APPRIS", href = "https://appris.bioinfo.cnio.es"), 
      " transcript is automatically selected (Fig. 1A)."),
    
    strong("Select plots"),
    p("If no pre-selection has been made, all plot types are displayed on the main panel directly after the query entry in the search bar
      including the Gene, Protein, ClinVar, gnomAD, and in-silico scores visualizations (see plot details below). The desired compilation
      of individual plots can be selected using check boxes under ", em("Select plots"), "(Fig. 1B)."),
    
    img(src = "www/Fig_aRgus_1.png", height = "400px"),
    p(strong("Fig. 1: "), em("Select gene "), "(A) and ", em("Select plots "), "(B) functions."),
    br(),
    
    strong("ClinVar significance"),
    p("Here, the display of all ClinVar variants is selected by default. Alternatively, the ClinVar plot can be restricted to ", em("pathogenic "), 
      "(classified as pathogenic or likely pathogenic) or ", em("benign "), "variants (benign, likely benign, uncertain significance, conflicting
      interpretations) (Fig. 2)."),
    
    img(src = "www/Fig_aRgus_2.png", height = "300px"),
    p(strong("Fig. 2: "), "Selection of both pathogenic and benign (A) or only pathogenic ClinVar variants (B) using the ", em("ClinVar significance "), "function."),
    br(),
    
    strong("Select scores"),
    p("Under ", em("Select scores"), ", a personal synopsis of ", em("in-silico "), "variant effect prediction scores can be compiled. Up to three scores can be
      plotted simultaneously in an arbitrary order. Additionally, values of all scores, available on aRgus, are listed in the tab ", em("In-silico scores "), 
      "of the bottom table (see section ", em("Table"), ")."),
    
    img(src = "www/Fig_aRgus_3.png", height = "200px"),
    p(strong("Fig. 3: "), "Score selection using the drop-down list (A) for simultaneous plotting of up to three different scores (B)."),
    br(),
    
    strong("Figure download options"),
    p("Each plot generated by aRgus can be exported separately. Here, the desired file format can be set to PNG or SVG (", em("Download format"), 
      "). Below, the plot dimensions can be adjusted by ", em("Width"), " and ", em("Height"), " (unit = inches)."),
    
    h3("Plots"),
    
    p("The desired compilation of plots can be set using the checkboxes under ", em("Select plots"), " (see ", em("Toolbar"), "). All plots 
      can be individually exported as PNG or SVG files using the download buttons on the top right of each plot (see ", em("Toolbar"), " > ", 
      em("Figure download options"), ")."),
    
    strong("Transcript"),
    p("In this plot, the canonical, unspliced transcript is shown as linear representation according to its chromosomal coordinates (x-axis) with the
      untranslated regions as gray backbone and the exons as green rectangles from left to right in ascending order regardless of the gene's position on  
      the genomic forward or reverse strand. The underlying Ensembl and RefSeq transcript identifiers are depicted in the title. The chromosomal positions
      of pathogenic and benign ClinVar variants are indicated as lollipop segments on top. Variant descriptions to corresponding lollipops can be faded-in 
      by selection in the ClinVar table (see ", em("Table"), " > ", em("ClinVar"), ") (Fig. 4)."),
    
    img(src = "www/ASS1_transcript_2021-12-19-1.png", height = "300px"),
    p(strong("Fig. 4: "), "The ", em("Transcript "), "plot with selection of a pathogenic variant selected in the ", em("ClinVar"), " table (red)."),
    br(),
    
    strong("Protein"),
    p("Here, the translated protein is shown in a linear representation including amino acid positions on the x-axis. Protein domains and regions 
      annotated in the UniProt database are shown as colored rectangles and depicted in the figure legend. The corresponding UniProt identifier is 
      displayed in the plot title. As in the Gene plot, amino acid positions of pathogenic and benign ClinVar variants are indicated as lollipop 
      segments and variant description can be faded-in by selection in the ", em("ClinVar"), " table (Fig. 5)."),
    img(src = "www/ASS1_protein_2021-12-19.png", height = "300px"),
    p(strong("Fig. 5: "), "The ", em("Protein "), "plot with selection of a pathogenic variant selected in the ", em("ClinVar "), "table (red). "),
    br(),
    
    strong("ClinVar"),
    p("In this plot, pathogenic and benign Clinvar variants are color-coded in red and blue, respectively, and can be either plotted together or
      solely according to their clinical significance using the ", em("ClinVar significance "), "selection bar (see Toolbar > ClinVar significance). 
      At the bottom, the amino acid position of the respective variants is shown as colored segments. On top, a density plot displays the variant 
      distribution with respect to their variant count as depicted in the ClinVar database (see Fig. 2)."),
    
    strong("gnomAD"),
    p("This plot displays variant distributions and frequencies of healthy individuals gathered in the comprehensive gnomAD database. 
      gnomAD variants are shown as bars according to their amino acid position and their allele count with logarithmic (log10) scaling of the y-axis.
      Variants are grouped in two plots depending on their identification within the scope of clinical testing using whole-exome (top, green) or whole-genome 
      sequencing (bottom, blue; Fig. 6)."),
    p("Remark: Regarding genetic disorders with recessive inheritance, pathogenic variants are naturally present in healthy heterozygous variant carriers
      within the general population and therefore not be generally classified as \"benign\"."),
    
    img(src = "www/ASS1_GnomAD.png", height = "300px"),
    p(strong("Fig. 6: "), "The ", em("gnomAD "), "plot with variants from the exomes (green) and genomes (blue) datasets."),
    br(),
    
    strong(em("In-silico "), "scores"),
    p("This plot visualizes value distributions of ", em("in-silico"), " scores that predict the variant effect on structure and function 
      of the translated protein. The score values are pre-calculated for all biologically possible non-synonymous single nucleotide variants (nsSNVs)
      including three possible nucleotide substitutions at any base position throughout the transcript. Depending on the position in the base triplet, 
      up to nine different resulting amino acid substitutions are possible, in case of a missense variant, from which any substitution is assigned by a
      distinct ", em("in-silico"), " score value. These detailed scores are listed in the ", em("In-silico scores"), " table (see ", em("Table"), " > ", 
      em("in-silico"), " scores). For intuitive interpretation of these tabular data, this plot comprises two different visualizations of the ", 
      em("in-silico"), " score value distributions throughout the amino acid sequence. For both approaches, arithmetic means of the different score
      values at any amino acid position are calculated. On top, the resulting data is shown as a smoothed curve based on a polynomial regression model. 
      At the bottom, the data is visualized as a heatmap where score values are color-coded with respect to their prediction as a rather damaging or 
      non-damaging effect on the protein. The score range is shown in the legend below and color-coded with respect to the recommended cut-off value. 
      Using the ", em("Score selection"), " bar, the desired score can be chosen from a list of 26 different scores and up to three scores can be
      displayed simultaneously for comparison (see ", em("Scores"), "). By default, the REVEL and CADD_phred scores are shown initially."),
    
    img(src = "www/ASS1_InSilicoScore_REVEL.png", height = "300px"),
    p(strong("Fig. 7: "), "The ", em("in-silico scores"), " plot with two different visualizations including a smoothed curve (top) and heatmap (bottom)."), 
    br(),
    
    strong("Score statistics"),
    p("Here, score values are assigned to the corresponding pathogenic and benign ClinVar as well as gnomAD variants and are compared to 
      score values of all biologically possible nsSNV shown in the ", em("in-silico"), " table using a t-test. These groupwise comparisons are visualized as violin 
      plots with included boxplots indicating median as well as first and third quartile (Fig. 8). Significant differences are indicated with * (p-value < 0.05), 
      ** (p-value < 0.01), and *** (p-value < 0.001). Thereby, a possible discrimination between pathogenic and benign variants by a specific score can be analyzed."),
    
    img(src = "www/ASS1_ScoreStatistics.png", height = "400px"),
    p(strong("Fig. 8: "), "The ", em("Score statistics"), "plot statistically comparing score values of pathogenic and benign ClinVar as well as gnomAD 
      variants with all biologically possible nsSNVs."),
    br(),
    
    h3("Tables"),
    
    p("The bottom table can be completely displayed by scrolling down. Data presented in the table can be filtered using the filter bars on top 
      of each column. The number of displayed rows can be chosen from 25 to 100 using the Show [\u2026] entries bar on the bottom left. Additionally, 
      specific terms in the table can be searched using the search bar on the top right. Data shown in both tabs can be copied to the clipboard 
      and exported as CSV and Excel files using the buttons ", em("Copy"), ", ", em("CSV"), ", and ", em("Excel"), "in the top part of the 
      table (Fig. 8)."),
    
    img(src = "www/Tbl1.png", height = "300px"),
    br(),
    
    strong("ClinVar"),
    p("Here, all ClinVar variants of the queried gene are listed including their position specification related to chromosome and transcript as 
      well as the associated phenotype as specified in the ClinVar entries. Compilation of ClinVar variant details was performed using the Simple 
      ClinVar filtering algorithm. If the user is interested in a specific variant or a subset of variants, single or multiple rows can be selected 
      and subsequently, the respective variants are displayed in the ", em("Gene"), " and ", em("Protein"), " plots with highlighting according 
      to their clinical significance (red = pathogenic, blue = benign). ClinVar variant distribution data are additionally visualized in the ",
      em("ClinVar"), " plot (see ", em("Plots"), " > ", em("ClinVar"), ")."),
    
    strong("In-silico scores"),
    p("In this tab, all biologically possible nsSNVs are listed including their corresponding ", em("in-silico"), " prediction score values 
      (see ", em("Plots"), " > ", em("In-silico scores"), "). If the simulated variant is present in the gnomAD exomes or genomes database, the 
      respective allele count and frequency are shown in the last four columns ", em("gnomAD_exomes_AC, gnomAD_exomes_AF, gnomAD_genomes_AC"), 
      ", and ", em("gnomAD_genomes_AF"), ". These tabular data are additionally visualized in the gnomAD and ", em("in-silico"), " scores plots 
      (see ", em("Plots"), ". Table cells containing score values are colored in orange if a damaging variant effect is predicted according
      to the recommended value ranges of the score creators. As the ", em("MVP"), " score provides two different cut-off values for \"constraint\" and
      \"non-constraint genes\", cells are color-coded in red and orange, respectively. Additional to the search and filter functions, score ranges 
      can be selected using a scrollbar under the column title."),
    
    h2("Scores"), 
    p("The subset of 43 pre-calculated variant effect prediction scores provided on aRgus is derived from dbNSFP which is an integrated database of
      functional annotations from multiple sources for the comprehensive collection of 84,013,490 human nsSNVs. A list of the scores found on aRgus 
      with their corresponding source is shown below. The first 26 scores can additionally be visualized using the In silico scores and Score statistics 
      plots and all 43 scores can be found in the in silico scores table. For further details on score ranges please refer to the ", 
      a("dbNSFP v.4.1a Readme", href = "https://drive.google.com/file/d/1VINShZQYN63HvigJA_SiuOYIk8TWl5Fh/view?usp=sharing"), " file, ", 
      a("dbNSFP homepage", href = "https://sites.google.com/site/jpopgen/dbNSFP"), ", and the score web references listed below:"),
    
    tags$ol(
      tags$li(strong("SIFT:"), a("https://sift.bii.a-star.edu.sg/www/history.html", href = "https://sift.bii.a-star.edu.sg/www/history.html")),
      tags$li(strong("SIFT4G: "), a("http://sift.bii.a-star.edu.sg/sift4g/public//Homo_sapiens/", href = "http://sift.bii.a-star.edu.sg/sift4g/public//Homo_sapiens/")),
      tags$li(strong("Polyphen2_HDIV: "), a("http://genetics.bwh.harvard.edu/pph2/", href = "http://genetics.bwh.harvard.edu/pph2/")),
      tags$li(strong("Polyphen2_HVAR: "), a("http://genetics.bwh.harvard.edu/pph2/", href = "http://genetics.bwh.harvard.edu/pph2/")),
      tags$li(strong("LRT: "), a("http://www.genetics.wustl.edu/jflab/lrt_query.html", href = "http://www.genetics.wustl.edu/jflab/lrt_query.html")),
      tags$li(strong("MutationTaster: "), a("http://www.mutationtaster.org/", href = "http://www.mutationtaster.org/")),
      tags$li(strong("MutationAssessor: "), a("http://mutationassessor.org/", href = "http://mutationassessor.org/")),
      tags$li(strong("FATHMM: "), a("http://fathmm.biocompute.org.uk", href = "http://fathmm.biocompute.org.uk")),
      tags$li(strong("PROVEAN: "), a("http://provean.jcvi.org/index.php", href = "http://provean.jcvi.org/index.php")),
      tags$li(strong("VEST4: "), a("http://karchinlab.org/apps/appVest.html", href = "http://karchinlab.org/apps/appVest.html")),
      tags$li(strong("MetaSVM: "), a("doi: 10.1093/hmg/ddu733", href = "doi: 10.1093/hmg/ddu733")),
      tags$li(strong("MetaLR: "), a("doi: 10.1093/hmg/ddu733", href = "doi: 10.1093/hmg/ddu733")),
      tags$li(strong("M-CAP: "), a("http://bejerano.stanford.edu/MCAP/", href = "http://bejerano.stanford.edu/MCAP/")),
      tags$li(strong("REVEL: "), a("https://sites.google.com/site/revelgenomics/", href = "https://sites.google.com/site/revelgenomics/")),
      tags$li(strong("MutPred: "), a("http://mutpred.mutdb.org/", href = "http://mutpred.mutdb.org/")),
      tags$li(strong("MVP: "), a("https://github.com/ShenLab/missense", href = "https://github.com/ShenLab/missense")),
      tags$li(strong("MPC: "), a("ftp://ftp.broadinstitute.org/pub/ExAC_release/release1/regional_missense_constraint/", href = "ftp://ftp.broadinstitute.org/pub/ExAC_release/release1/regional_missense_constraint/")),
      tags$li(strong("PrimateAI: "), a("https://github.com/Illumina/PrimateAI", href = "https://github.com/Illumina/PrimateAI")),
      tags$li(strong("DEOGEN2: "), a("https://deogen2.mutaframe.com/", href = "https://deogen2.mutaframe.com/")),
      tags$li(strong("BayesDel_addAF: "), a("http://fengbj-laboratory.org/BayesDel/BayesDel.html", href = "http://fengbj-laboratory.org/BayesDel/BayesDel.html")),
      tags$li(strong("BayesDel_noAF: "), a("http://fengbj-laboratory.org/BayesDel/BayesDel.html", href = "http://fengbj-laboratory.org/BayesDel/BayesDel.html")),
      tags$li(strong("ClinPred: "), a("https://sites.google.com/site/clinpred/home", href = "https://sites.google.com/site/clinpred/home")),
      tags$li(strong("LIST-S2: "), a("https://precomputed.list-s2.msl.ubc.ca/", href = "https://precomputed.list-s2.msl.ubc.ca/")),
      tags$li(strong("CADD_phred: "), a("http://cadd.gs.washington.edu/", href = "http://cadd.gs.washington.edu/")),
      tags$li(strong("DANN: "), a("https://cbcl.ics.uci.edu/public_data/DANN/", href = "https://cbcl.ics.uci.edu/public_data/DANN/")),
      tags$li(strong("fathmm-MKL_coding: "), a("http://fathmm.biocompute.org.uk/fathmmMKL.htm", href = "http://fathmm.biocompute.org.uk/fathmmMKL.htm")),
      tags$li(strong("fathmm-XF_coding: "), a("http://fathmm.biocompute.org.uk/fathmm-xf/", href = "http://fathmm.biocompute.org.uk/fathmm-xf/")),
      tags$li(strong("Eigen-raw_coding: "), a("http://www.columbia.edu/~ii2135/eigen.html", href = "http://www.columbia.edu/~ii2135/eigen.html")),
      tags$li(strong("Eigen-PC-raw_coding: "), a("http://www.columbia.edu/~ii2135/eigen.html", href = "http://www.columbia.edu/~ii2135/eigen.html")),
      tags$li(strong("GenoCanyon: "), a("http://genocanyon.med.yale.edu/index.html", href = "http://genocanyon.med.yale.edu/index.html")),
      tags$li(strong("integrated_fitCons: "), a("http://compgen.bscb.cornell.edu/fitCons/", href = "http://compgen.bscb.cornell.edu/fitCons/")),
      tags$li(strong("GM12878_fitCons: "), a("http://compgen.bscb.cornell.edu/fitCons/", href = "http://compgen.bscb.cornell.edu/fitCons/")),
      tags$li(strong("H1.hESC_fitCons: "), a("http://compgen.bscb.cornell.edu/fitCons/", href = "http://compgen.bscb.cornell.edu/fitCons/")),
      tags$li(strong("HUVEC_fitCons: "), a("http://compgen.bscb.cornell.edu/fitCons/", href = "http://compgen.bscb.cornell.edu/fitCons/")),
      tags$li(strong("LINSIGHT: "), a("http://compgen.cshl.edu/~yihuang/LINSIGHT/", href = "http://compgen.cshl.edu/~yihuang/LINSIGHT/")),
      tags$li(strong("GERP++_RS: "), a("http://mendel.stanford.edu/SidowLab/downloads/gerp/", href = "http://mendel.stanford.edu/SidowLab/downloads/gerp/")),
      tags$li(strong("phyloP100way_vertebrate: "), a("http://hgdownload.soe.ucsc.edu/goldenPath/hg38/phyloP100way/", href = "http://hgdownload.soe.ucsc.edu/goldenPath/hg38/phyloP100way/")),
      tags$li(strong("phyloP30way_mammalian: "), a("http://hgdownload.soe.ucsc.edu/goldenPath/hg38/phyloP30way/", href = "http://hgdownload.soe.ucsc.edu/goldenPath/hg38/phyloP30way/")),
      tags$li(strong("phyloP17way_primate: "), a("http://hgdownload.soe.ucsc.edu/goldenPath/hg38/phyloP17way/", href = "http://hgdownload.soe.ucsc.edu/goldenPath/hg38/phyloP17way/")),
      tags$li(strong("phastCons100way_vertebrate: "), a("http://hgdownload.soe.ucsc.edu/goldenPath/hg38/phastCons100way/", href = "http://hgdownload.soe.ucsc.edu/goldenPath/hg38/phastCons100way/")),
      tags$li(strong("phastCons30way_mammalian: "), a("http://hgdownload.soe.ucsc.edu/goldenPath/hg38/phastCons30way/", href = "http://hgdownload.soe.ucsc.edu/goldenPath/hg38/phastCons30way/")),
      tags$li(strong("phastCons17way_primate: "), a("http://hgdownload.soe.ucsc.edu/goldenPath/hg38/phastCons17way/", href = "http://hgdownload.soe.ucsc.edu/goldenPath/hg38/phastCons17way/")),
      tags$li(strong("SiPhy_29way_logOdds: "), a("https://www.broadinstitute.org/mammals-models/29-mammals-project-supplementary-info", href = "https://www.broadinstitute.org/mammals-models/29-mammals-project-supplementary-info"))
    ),
    
    h2("Browser and OS compatibility"),
    p("aRgus was tested on different browsers and operating systems which are shown in the table below:"),
    p(strong("Linux: 5.15.10"), "Chrome (96.0.4664.110), Edge (n/a), Firefox (87.0), Safari (n/a)"),
    p(strong("MacOS: 12.0.1"), "Chrome (96.0.4664.110), Edge (96.0.1054.62), Firefox (95.0.1), Safari (15.1)"),
    p(strong("iOS: 15.1"), "Chrome (96.0.4664.110), Edge (96.1054.49), Firefox (40.1), Safari (15.0)"),
    p(strong("Windows: 10"), "Chrome (96.0.4664.110), Edge (96.0.1054.62), Firefox (95.0.1), Safari (n/a)")
    
    

  )
}
    

## To be copied in the UI
# mod_help_ui("help_ui_1")
    